linter: add synchronization detection to loopvarcapture

renatolabs opened the pull request:
This adds a number of improvements to the `loopvarcapture` linter introduced
in #82324. Most of the improvements are related to the linter being more aware
of explicit synchronization the developer has added that makes references to
loop variables in goroutines safe (examples: `sync.WaitGroup`, channels, etc).

Previously, the `loopvarcapture` linter would flag any references to
loop variables found in closures passed to `go` or `defer`
statements. That happened even when the engineer intentionally
referenced the loop variable because extra synchronization would make
sure that a new iteration of the loop would not begin until the Go
routines were finished (by using wait groups, channels, or other
mechanisms). This led to safe code being reported, which can be
frustrating and add noise.

This PR adds the following changes to `loopvarcapture`:

* References synchronized using `sync.WaitGroup` are no longer flagged by the linter
(specifically, this happens if the Go routine calls `wg.Done()` *after* the reference
occurs, and there's a statement in the loop after `go` that calls `Wait()` on the same
wait group). The example below, for instance, is no longer invalid:

```go
for k, v := range coll {
    // ...
    var wg sync.WaitGroup
    // ...
    go func() {
        defer wg.Done()
        // ...
        someFunc(k, v)
        // ...
    }()
    // ...
    wg.Wait()
}
```

* Synchronization using channels is also supported. If the Go routine writes to
or closes a channel that is later read within the loop, the access is considered safe.

* Go-routine wrappers (such as `errgroup`, cockroach's `ctxgroup`, etc) generally
provide a `Wait()` function that blocks until all the goroutines created finish. The linter
is now able to understand these kinds of synchronization as well.

* Referencing a loop variable in the closure passed to `t.Run()` is safe by default (see next
item in this list). Even though the capture happens "by reference", the `Run()` function
will execute the closure immediately and will block until it returns. This means that references
to loop variables in that closure are safe. The semantics of `t.Run()` are now part of
the linter (since this is a very common case).

* Accessing loop variables in `t.Run()` is **not** safe if the test runs in parallel with other tests
(i.e., if there's a call to `t.Parallel()`). In such cases, referencing a loop variable in `t.Run()`is not
safe. The linter now finds invalid references of this kind, and #86488 was found because of that.

* Other invalid uses of loop variables that are now detected by the linter:

```go
func (s *S) Fn() {
    // ...
}

for _, n := range items {
    go fn(&n) // taking the address of a loop variable and passing it to a goroutine

    list := []S{/* ... */}
    for _, s := range list {
        go s.Fn() // will not call `Fn()` on every item in `list`
    }
}
```

Epic: None.

Release note: None.

Commit:
linter: add synchronization detection to `loopvarcapture`

Previously, the `loopvarcapture` linter would flag any references to
loop variables found in closures passed to `go` or `defer`
statements. That happened even when the engineer intentionally
referenced the loop variable because extra synchronization would make
sure that a new iteration of the loop would not begin until the Go
routines were finished (by using wait groups, channels, or other
mechanisms). This led to safe code being reported, which can be
frustrating and add noise.

This commit adds simple synchronization detection for the
`loopvarcapture` linter. With the changes introduced in this commit,
code of the following form will no longer be flagged:

```
for k, v := range coll {
    // ...
    var wg sync.WaitGroup
    // ...
    go func() {
        defer wg.Done()
        // ...
        someFunc(k, v)
        // ...
    }()
    // ...
    wg.Wait()
}
```

The previous code is safe because it the loop waits on the WaitGroup,
and the go routine calls `Done` on it. A similarly supported scenario
is go routines that `close` a channel that is subsequently read in the
body of the loop.

Release note: None.

Commit:
linter: recognize synchronization used by go routine wrappers.

This commit updates the `loopvarcapture` linter to recognize
synchronization when "go routine wrappers" are used. These wrappers
are either provided by the standard library (e.g., `errgroup`), or by
custom code in the cockroach repository (e.g., `ctxgroup`).

Go routine wrappers now have the option to specify if they provide a
"wait function"; when a go routine references a loop variable using a
go wrapper, but the body of the loop calls the wait function, then the
refernece is deemed safe since the loop variable will not change while
the Go routine is executing.

Release note: None.

Commit:
linter: better handling in the context of t.Run() calls

The `loopvarcapture` linter reports references to loop variables in Go
routines created in any expression in the body of a loop. A common
occurrence of such references, however, are in tests that iterate
over a set of test cases, and call `t.Run()` to run the test with a
specific input or configuration. Consider the following code:

```
for i, tc := range testCases {
    t.Run(fmt.Sprintf("run %d", i), func(t *testing.T) {
        // ...
        go func() {
            defer close(ch)
            // ...
            doSomething(tc)
        }()

        // ...
        <-ch
    })
}
```

Even with synchronization detection, the linter would report this as
an unsafe reference because it cannot know what the `Run()` function
will do with the closure passed as parameter: it could call
immediately, in a Go routine, or save internally for execution
later. Only the first scenario would be safe.

This commit adds support for the `loopvarcapture` to understand the
semantics of `t.Run()`. After this commit, the function above will no
longer lead to a report because it knows that the closure will run
immediately, and there's synchronization code in the closure that
guarantees the closure won't return before the Go routine is finished.

The linter also understands `t.Parallel()`: if the closure above were
to call `t.Parallel()`, than an issue would be reported about the
`doSomething()` call, since the loop variable reference would be
unsafe in that case.

Release note: None.

Commit:
linter: support method calls and address-of loop variables.

This commit extends the `loopvarcapture` linter to support two
problematic patterns.

The first is taking the address of a loop variable, and passing it as
parameter to a function invoked with the `go` keyword, as in the
example below:

```
for _, tc := range testCases {
    // ...
    go myFunc(1, &tc) // <- invalid reference
    // ...
}
```

The second is invoking a method defined on a pointer type on the loop
variable, as below:

```
for _, tc := range testCases {
    // ...
    go tc.run() // <- where run is defined on a pointer
    // ...
}
```

Occurrences of the two patterns outlined above will now be reported.

Release note: None.

Commit:
lint: consider position of synchronization call

This updates the `loopvarcapture` linter to take into account the
relative position in a closure of a synchronization call. Previously,
the following code would be considered safe:

```
for _, tc := range testCases {
    // ...
    ch := make(chan error)
    go func() {
        close(ch)
        runTest(tc)
    }()
    <-ch
}
```

However, the code above is not really safe because the `close` call
happens *before* the reference to the loop variable, allowing the loop
to make progress before the reference actually occurred.

This commit will now take into account the position of the
synchronization call. The code above will only be safe if the `close`
call happens after `runTest`, or if `close` is called within `defer`.

Release note: None.

Release justification:

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86836)
<!-- Reviewable:end -->

Commit:
lint: add async task functions from stopper package to loopvarcapture.

Release note: None

Commit:
lint: Go 1.19 updates

Release note: None

Commit:
lint: correctly handle struct fields in loopvarcapture

`go/parser` cannot accurately resolve composite literals, as it's not
possible to determine at parse time if the keys are identifiers or
field names. Turns out even the type information available as
`pass.TypesInfo` is accurate.

In this commit we add a workaround of keeping track of struct literals
and ignoring keys in key-value expressions to avoid false positives
where the linter flags a struct initialization as a reference to a
loop variable when the field name is the same as the loop variable.

Relevant code in the `go/parser` package:

https://cs.opensource.google/go/go/+/master:src/go/parser/parser.go;l=1602-1617;drc=0bd308ff27822378dc2db77d6dd0ad3c15ed2e08.

Release note: None.

Commit:
lint: document unsupported patterns in loopvarcapture

Release note: None

Commit:
github-pull-request-make: log query command

Release note: None

renatolabs commented:
This was useful and helped us find a couple of bugs in CockroachDB, and a number of bugs in other open source projects [1]. However, the complexity added here probably does not justify merging it on master, especially now that we expect similar functionality to be available in the Go compiler in the future [2].

Closing.

[1] https://docs.google.com/document/d/1sdyCFIhFez2I1phVgKRxshZBAgBX1uQAG9OiUSFSdUQ/edit?usp=sharing
[2] https://github.com/golang/go/issues/60078
